### PR TITLE
Refactor webhook to allow adding conversion support

### PR DESCRIPTION
**Pull request by dprotaso**
Testing changes with serving here: https://github.com/knative/serving/pull/6496

### CODE REVIEW

File: webhook/admission.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		var ttStart = time.Now()
+		logger.Infof("Webhook ServeHTTP request=%#v", r)
+
+		var review admissionv1beta1.AdmissionReview
+		if err := json.NewDecoder(r.Body).Decode(&review); err != nil {
+			http.Error(w, fmt.Sprintf("could not decode body: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		logger = logger.With(
```
Comment (vagababov) [inline webhook/admission.go]: Some of those are strings and don't require `Sprint`, e.g. subresource.

File: webhook/admission.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			zap.String(logkey.SubResource, fmt.Sprint(review.Request.SubResource)),
+			zap.String(logkey.UserInfo, fmt.Sprint(review.Request.UserInfo)))
+		ctx := logging.WithLogger(r.Context(), logger)
+
+		var response admissionv1beta1.AdmissionReview
+		reviewResponse := c.Admit(ctx, review.Request)
+		logger.Infof("AdmissionReview for %#v: %s/%s response=%#v",
+			review.Request.Kind, review.Request.Namespace, review.Request.Name, reviewResponse)
+
+		if !reviewResponse.Allowed {
```
Comment (vagababov) [inline webhook/admission.go]: `if` and `else if` do the same thing, so you should just extend the `if` condition

File: webhook/webhook.go [hunk 3]
```diff
@@ -111,35 +105,37 @@ func New(
...
-	logger := ac.Logger
+func (w *Webhook) Run(stop <-chan struct{}) error {
+	logger := w.Logger
 	ctx := logging.WithLogger(context.Background(), logger)
 
 	server := &http.Server{
-		Handler: ac,
-		Addr:    fmt.Sprintf(":%v", ac.Options.Port),
+		Handler: w,
+		Addr:    fmt.Sprintf(":%v", w.Options.Port),
```
Comment (vagababov) [inline webhook/webhook.go]:
```suggestion
		Addr:    fmt.Sprintf(":%d", w.Options.Port),
```

File: webhook/webhook.go [hunk 4]
```diff
@@ -111,35 +105,37 @@ func New(
...
-		admissionControllers: acs,
-		Logger:               logger,
-	}, nil
+	webhook.mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, fmt.Sprintf("no admission controller registered for: %s", r.URL.Path), http.StatusBadRequest)
+	})
+
+	for _, c := range controllers {
+		webhook.mux.Handle(
+			c.Path(),
```
Comment (mattmoor) [inline webhook/webhook.go]: You removed the checking for Path duplicates?

Comment (mattmoor) [reply webhook/webhook.go]: ah, maybe the library panics? (I see the recover above)

Comment (mattmoor) [reply webhook/webhook.go]: _... read the comments, Matt_ 🤦‍♂

**Pull request merged**